### PR TITLE
Add the ability to terraform planets to AI

### PR DESCRIPTION
**Pull request by rrobgill**
* Terraformer and gaia transformation can be built by AI after
  each relevant tech is researched

  Limits:

  - No more than one of each building may be queued per turn

  - Will be queued in turn on the planet with the largest
    population, and planet must already be at its target population

  - Will not be enqueued if enemy ships are present in system

* GRO_TERRAFORM and GRO_GAIA_TRANS added to Growth techs

Signed-off-by: Rob Gill <rrobgill@protonmail.com>

### CODE REVIEW

**Comment by geoffthemedio** [comment]
````
./AI/ProductionAI.py:1111:17: E129 visually indented line with same indent as next logical line
./AI/ProductionAI.py:1112:21: E117 over-indented
````

File: default/python/AI/ProductionAI.py [hunk 1]
```diff
@@ -1091,6 +1091,42 @@ def generate_production_orders():
...
             warning("Failed enqueueing %s at planet %s, got result %d", building_name, planet, res)
 
+    if tech_is_complete(GRO_GAIA_TRANS):
+        terraformers = ["BLD_GAIA_TRANS", "BLD_TERRAFORM"]
+    elif tech_is_complete(GRO_TERRAFORM):
+        terraformers = ["BLD_TERRAFORM"]
+    else:
+        terraformers = []
+    if terraformers:
+        queued_locs = [element.locationID for element in production_queue if (element.name in terraformers)]
```
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py]: Use set
```python
queued_locs = {element.locationID for element in production_queue if (element.name in terraformers)}
```

File: default/python/AI/ProductionAI.py [hunk 2]
```diff
@@ -1091,6 +1091,42 @@ def generate_production_orders():
...
+            planet = universe.getPlanet(pid)
+            if (planet.type not in [fo.planetType.gasGiant, fo.planetType.asteroids]
+                and planet.currentMeterValue(fo.meterType.population) == planet.currentMeterValue(fo.meterType.targetPopulation)
+                and pid not in queued_locs
+                and aistate.systemStatus.get(planet.systemID, {}).get('fleetThreat', 1) == 0
+                and "GAIA_SPECIAL" not in planet.specials
+                and "NO_TERRAFORM" not in fo.getSpecies(planet.speciesName).tags
+                and not [bldg for bldg in planet.buildingIDs if universe.getBuilding(bldg).buildingTypeName in terraformers]):
+                    candidates.append([pid, planet.currentMeterValue(fo.meterType.population)])
+        if candidates:
```
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py]: Lets remove `if`  

It is ok to iterate over empty list. 

You still can use else:  https://book.pythontips.com/en/latest/for_-_else.html#for-else

File: default/python/AI/ProductionAI.py [hunk 3]
```diff
@@ -1091,6 +1091,42 @@ def generate_production_orders():
...
+            candidates.sort(key=itemgetter(1), reverse=True)
```
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py]: Switch order of values on candidates and you can omit `key` argument.

File: default/python/AI/ProductionAI.py [hunk 4]
```diff
@@ -1091,6 +1091,42 @@ def generate_production_orders():
...
+        for pid in state.get_inhabited_planets():
+            planet = universe.getPlanet(pid)
+            if (planet.type not in [fo.planetType.gasGiant, fo.planetType.asteroids]
+                and planet.currentMeterValue(fo.meterType.population) == planet.currentMeterValue(fo.meterType.targetPopulation)
+                and pid not in queued_locs
+                and aistate.systemStatus.get(planet.systemID, {}).get('fleetThreat', 1) == 0
+                and "GAIA_SPECIAL" not in planet.specials
+                and "NO_TERRAFORM" not in fo.getSpecies(planet.speciesName).tags
+                and not [bldg for bldg in planet.buildingIDs if universe.getBuilding(bldg).buildingTypeName in terraformers]):
+                    candidates.append([pid, planet.currentMeterValue(fo.meterType.population)])
```
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py]: Please add tuples, not lists.

File: default/python/AI/ProductionAI.py [hunk 5]
```diff
@@ -1091,6 +1091,42 @@ def generate_production_orders():
...
+        if candidates:
+            candidates.sort(key=itemgetter(1), reverse=True)
+            for bldg in terraformers:
+                for pid in candidates:
```
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py]: it is not pid, it is a pair of pid and population.  

You can use unpacking:  `for pid, _ in candidates: planet = universe.getPlanet(pid)`

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
I am not sure how it will affect AI behavior. 
As I recall this is a huge building. We don't evaluate the benefits of them.  Is they always worth building?

**Comment by geoffthemedio** [comment]
@rrobgill ?

**Comment by rrobgill** [comment]
no plans.

**Pull request closed without merging**